certbot-dns-google/tests/dns_google_test.py

```diff
@@ -86,9 +86,19 @@ def _setUp_client_with_mock(self, zone_request_side_effect):
         mock_mz.list.return_value.execute.side_effect = zone_request_side_effect
 
         mock_rrs = mock.MagicMock()
-        rrsets = {"rrsets": [{"name": "_acme-challenge.example.org.", "type": "TXT",
+        def rrs_list(project=None, managedZone=None, name="foo", type="TXT"):
+            response = {"rrsets": []}
+            if name == "_acme-challenge.example.org.":
+                response = {"rrsets": [{"name": "_acme-challenge.example.org.", "type": "TXT",
                               "rrdatas": ["\"example-txt-contents\""]}]}
-        mock_rrs.list.return_value.execute.return_value = rrsets
+            class x:
+                @staticmethod
+                def execute():
+                    if rrs_list_error is not None:
+                        raise rrs_list_error
+                    return response
+            return x
```

Comment:
Yeah, I also think is a bit weird. I'll come up with something simpler 👍 . 

The motivation behind was that while testing `test_get_existing` (I wanted to have a mock with the following requirements on the `execute` function:
- it can return a different response depending on what's given to the `get_existing_txt_rrset` method, i.e.: first case calling `get_existing_txt_rrset` with a record that exists, second case calling `get_existing_txt_rrset` that doesn't exist.
- it can return an exception/error if specified on the `_setUp_client_with_mock` function. 

https://github.com/certbot/certbot/blob/38893115572ee1bc91fcaf10669a80686795ba0b/certbot-dns-google/tests/dns_google_test.py#L290-L297

But I agree that I had to stretch it out a bit and the result is a bit complex for the reader.

Reply:
I may have been unclear, sorry. I really liked that you made that `rrs_list` mock function which varied its result by input. I think that this was the correct approach.

My suggestion was, rather than implementing the functionality using a `class x` implementing `execute`, to instead use a mock object to do the same thing, e.g.:

```python
        mock_rrs = mock.MagicMock()
        def rrs_list(project=None, managedZone=None, name=None, type=None):
            response = {"rrsets": []}
            if name == "_acme-challenge.example.org.":
                response = {"rrsets": [{"name": "_acme-challenge.example.org.", "type": "TXT",
                              "rrdatas": ["\"example-txt-contents\""]}]}
            mock_return = mock.MagicMock()
            mock_return.execute.return_value = response
            mock_return.execute.side_effect = rrs_list_side_effect
            return mock_return
        mock_rrs.list.side_effect = rrs_list
```

In fact getting rid of `rrs_list` caused a minor drop in test coverage, so it would be good to go back to the previous approach (without `class x`).

Reply:
ohh ok, no worries. I didn't the know the coverage went down. I just reverted the previous commit and push this suggestion you are making. Cool, this looks way better, I still need to learn a bit more about mocking in python 🙈 .